### ISSUE TITLE
When using disposable, typescript should never report the variable as unused

### ISSUE THREAD

**Issue by KristjanTammekivi** [microsoft/TypeScript#55538]
### 🔍 Search Terms

"using", "dispose", "asyncDispose", "is declared but its value is never read"

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

TypeScript reports variables declared via `using` as unused if they aren't used explicitly, but I would argue that they are used implicitly and shouldn't be reported as unused in any case.

### 📃 Motivating Example

[TypeScript Playground link](https://www.typescriptlang.org/play?noUnusedLocals=true&target=9&lib=lib.es2022.d.ts%2Clib.esnext.disposable.d.ts&ts=5.2.2#code/PTAEAEBsEsCMC5QFMDOAmADGtAaVA7JADwBcA6AE2hQAcB7FAQ1kiQCg2BjO-FE0AK4okAWQElioALyhGKAJ75OoABQBKaQD5QAbzagDoEKG69+kOpwDW02QHdG0fgHMkJADKWr6sr-2GAJzcBAPxdf0NDAG0AZXkAW1g6SDI5RU4AEWp6YQBdRDSlVQ0pbT1IisNjRgcnUAtrMiDWOSR1AG4IioBfLt7erh4+UHjHMJlC5XUtcMiax34haHxnEfFJCdrF4TEJIg6u+bqlldAAfXj1olsj7dErg96gA)

### 💻 Use Cases

1. What do you want to use this for?
2. What shortcomings exist with current approaches?
3. What workarounds are you using in the meantime?

Sometimes I just want the `using` for creating a mutex lock, and I don't need to interact with the variable in any way.

Workaround is to prefix the variable with _

**Comment by MartinJohns** [comment]
I would argue that it should be reported as unused. I believe in the majority of cases you actually want to use the resource for something, and not only treat it as an automatic scope. For those cases prefixing with an underscore is a good workaround.

Instead a proposal should be made to TC39 to support `using` declarations that don't bind any variable.

**Comment by andrewbranch** [comment]
I agree; I think it looks wrong enough that the `_` prefix, if it was really intended to be unused by explicit code, is really valuable.

**Comment by KristjanTammekivi** [comment]
https://github.com/tc39/proposal-discard-binding
I guess I have to wait for this.

**Comment by matthieusieben** [comment]
With the `using` becoming more and more widely supported, this issue should be re-considered. We are converting our code to use disposable. However, some of our dependencies have not been converted yet. Se we end-up doing:

```ts
const resource = getSomeOldResourceThatIsNotDisposable()
using resourceDisposer = { [Symbol.dispose]: () => resource.destroy() }
```

Both `https://github.com/tc39/proposal-discard-binding` and `DisposableStack` would allow to avoid doing this. However, neither are widely available today. And using `DisposableStack` for this purpose only requires to create an additional object (even if only one cleanup needs to happen).

I also believe that the error message is misleading. Indeed, those variables **are** used. Just not explicitly.

**Comment by snarbles2** [comment]
The error message I'm seeing is `'variable' is declared but its value is never read`, which I would say is reasonably accurate and clear.

I can't see how it makes sense to do anything here when there is both the void-binding proposal in process *and*, until then, the simple workaround of prefixing the identifier with an underscore.
